src/state/pbridge/providers/rango.ts

```diff
@@ -0,0 +1,140 @@
+import { parseUnits } from '@ethersproject/units';
+import { Currency, RANGO, Token, TokenAmount } from '@pangolindex/sdk';
+import {
+  BlockchainMeta,
+  EvmBlockchainMeta,
+  QuoteRequest,
+  QuoteResponse,
+  RangoClient,
+  SwapRequest,
+  SwapResponse,
+  TransactionType,
+} from 'rango-sdk-basic';
+import { NATIVE, RANGO_API_KEY, ZERO_ADDRESS } from 'src/constants';
+import { calculateTransactionTime } from 'src/utils';
+import { BridgePrioritizations, GetRoutes, GetRoutesProps, Route, Step } from '../types';
+
+export const getRangoRoutes: GetRoutes = async ({
+  amount,
+  slipLimit,
+  fromChain,
+  toChain,
+  fromAddress,
+  fromCurrency,
+  toCurrency,
+  recipient,
+}: GetRoutesProps) => {
+  const parsedAmount = parseUnits(amount, fromCurrency?.decimals).toString();
+  if (!fromCurrency?.symbol || !toCurrency?.symbol) {
+    return [];
+  }
+  const rango = new RangoClient(RANGO_API_KEY);
+  const chains = await rango.chains();
+  const isEvmBlockchain = (chain: BlockchainMeta): chain is EvmBlockchainMeta => chain.type === TransactionType.EVM;
+  const evmChainIdToName = Object.assign(
+    {},
+    ...chains.filter(isEvmBlockchain).map((chain) => ({
+      [chain.chainId.toString()]: chain.name,
+    })),
+  );
+
+  const fromCurrencyAddress =
+    !fromCurrency.address || [ZERO_ADDRESS, NATIVE].includes(fromCurrency.address) ? null : fromCurrency.address;
+  const toCurrencyAddress =
+    !toCurrency.address || [ZERO_ADDRESS, NATIVE].includes(toCurrency.address) ? null : toCurrency.address;
+
+  let request: QuoteRequest | SwapRequest = {
+    from: {
+      blockchain: evmChainIdToName[fromCurrency.chainId.toString()],
+      symbol: fromCurrency.symbol,
+      address: fromCurrencyAddress,
+    },
+    to: {
+      blockchain: evmChainIdToName[toCurrency.chainId.toString()],
+      symbol: toCurrency.symbol,
+      address: toCurrencyAddress,
+    },
+    amount: parsedAmount,
+  };
+  let rangoRouteRes: QuoteResponse | SwapResponse;
+
+  if (!fromAddress) {
+    rangoRouteRes = await rango.quote(request);
+  } else {
+    request = {
+      ...request,
+      fromAddress: fromAddress,
+      toAddress: recipient || fromAddress,
+      slippage: slipLimit,
+      disableEstimate: false,
+      referrerAddress: null,
+      referrerFee: null,
+    };
+    rangoRouteRes = await rango.swap(request);
+  }
+
+  // no routes found
+  if (rangoRouteRes?.resultType !== 'OK' || !rangoRouteRes?.route) return [];
+
+  // routes founds but failed to create tx or pass validation checks
+  if (rangoRouteRes?.resultType === 'OK' && !!rangoRouteRes?.error && !rangoRouteRes?.error?.includes('balance'))
+    return [];
+
+  const rangoRoute: Route | undefined = rangoRouteRes.route
```

Comment:
Other providers are not considering user balance when they want to show possible routes for the user and throw an error in executing routes regarding the insufficient balance. I put this condition here so that Rango will behave like other providers. (show possible routes but throw error in executing route)